jq255.md

```diff
@@ -0,0 +1,1425 @@
+
+The decoding process rebuilds either the original `P`, or the other
+representant `P+N` of the same group element:
+
+```python
+def field_to_group(v):
+    t = v**2
+    delta = (a**2 - 4*b)*(t**2) - 2*a*t + 1
+    E = sqrt(delta)       # Note: sqrt() returns the non-negative root
+    if E is False:
+        return False
+    Z = GFq(1)
+    U = v
+    T = t
+    return (E, Z, U, T)
+```
+
+The `field_to_group(v)` function returns a representant of the (unique)
+group element that encodes to the value `v`, or `False` if there is no
+such element. Take care that `sqrt()` MUST return the non-negative root.
+
+**Note:** The decoding process naturally returns a point in (extended)
+affine coordinates (`Z = 1`).
+
+### Point Addition
+
+The sum of two group elements is computed by adding the respective
+representants of the elements on the curve. In EZUT coordinates, the
+points `P1` and `P2` are added as follows:
+
+```python
+def point_add(P1, P2):
+    (E1, Z1, U1, T1) = P1
+    (E2, Z2, U2, T2) = P2
+    ap = -2*a              # The constant a'
+    bp = a^2 - 4*b         # The constant b'
+    e1e2 = E1*E2
+    z1z2 = Z1*Z2
+    u1u2 = U1*U2
+    t1t2 = T1*T2
+    tz = (Z1 + T1)*(Z2 + T2) - z1z2 - t1t2
+    eu = (E1 + U1)*(E2 + U2) - e1e2 - u1u2
+    hd = z1z2 - bp*t1t2
+    E3 = (z1z2 + bp*t1t2)*(e1e2 + ap*u1u2) + 2*bp*u1u2*tz
+    Z3 = hd**2
+    T3 = eu**2
+    U3 = half((hd + eu)**2 - Z3 - T3)   # Alternatively: U3 = hd*eu
+    return (E3, Z3, U3, T3)
+```
+
+If the point `P2` is in extended affine coordinates, i.e. such that `Z2`
+is statically known to be equal to 1, then the multiplication `Z1*Z2`
+can be avoided, and the computation of `tz` simplifies to `Z1*T2 + T1`.
+This specific case is known as a *mixed addition* and is typically
+encountered when multiplying the conventional generator by a scalar,
+using precomputed tables of small multiples of the generator.
+
+### Point Negation and Subtraction
+
+Given the point `P = (E:Z:U:T)`, its opposite is obtained as
+`-P = (E:Z:-U:T)`, i.e. simply negating the `U` coordinate in the field.
+The subtraction of `P2` from `P1` can be obtained by adding `-P2` to
+`P1`.
+
+```python
+def point_neg(P):
+    (E, Z, U, T) = P
+    return (E, Z, -U, T)
+
+def point_sub(P1, P2):
+    return point_add(P1, point_neg(P2))
+```
+
+### Point Doublings
+
+Suppose that we have `k >= 1` successive doublings to perform on an
+input point `P` (i.e. we want `(2^k)*P`). The general strategy is
+the following:
+
+  - Perform the first doubling with formulas that receive `P` in
+    EZUT coordinates, and output `2*P` in XWJ coordinates.
+  - Perform the other `k-1` doublings in XWJ coordinates.
+  - Convert the final result from XWJ to EZUT.
+
+Depending on the used curve, it can be beneficial to combine some of
+these doublings with a switch to the other representant (i.e. computing
+`2*P+N` instead of `2*P`); this does not change the output as a group
```

Comment:
```suggestion
`2*(P+N)` instead of `2*P`); this does not change the output as a group
```

Reply:
It's really `(2*P)+N` that I meant here. But otherwise yes, as an implementer you have the choice between `2*P`, `2*(P+N)`, `(2*P)+N` and `(2*(P+N))+N`, which all lead to the same group element. In XWJ coordinates, the switch from `P` to `P+N` means negating `w`, and replacing `x` with `b/x`; the latter change moves expressions from numerator to denominator and vice versa, and thus leads to different formulas, which may be better or worse from a performance point of view.

Indeed, for jq255e, the best performance seems to be obtained by computing `2*P`, while on jq255s it's better to use `(2*P)+N` (it reduces per-doubling cost from 1M+6S to 2M+4S).

Reply:
I have expanded the wording a bit to make things clearer.